### PR TITLE
Invite Code support for Private Rooms

### PR DESCRIPTION
**Pull request by analogrelay**
Allow room owners to manage invite codes that users can use to join rooms without requiring owners to directly add users.

@davidfowl: Ask and ye shall receive ;P

### CODE REVIEW

File: JabbR/Chat.js [hunk 1]
```diff
@@ -332,6 +332,11 @@
         ui.addMessage('Your gravatar has been set', 'notification', this.activeRoom);
     };
 
+    // Called when the server sends a notification message
+    chat.postNotification = function (msg, room) {
+        ui.addMessage(msg, 'notification', room);
+    }
```
Comment (davidfowl) [inline JabbR/Chat.js]: Missing semicolon.

File: JabbR/Commands/CommandManager.cs [hunk 2]
```diff
@@ -221,6 +232,25 @@ private bool TryHandleUserCommand(string commandName, string[] parts)
...
             return false;
         }
+
+        private void HandleInviteCode(ChatUser user, ChatRoom room, bool forceReset)
+        {
+            if (String.IsNullOrEmpty(room.InviteCode) || forceReset)
+            {
+                // Generate a new invite code
+                string code;
+                using (var crypto = new RNGCryptoServiceProvider())
```
Comment (davidfowl) [inline JabbR/Commands/CommandManager.cs]: Move this into a helper method and git it a meaningful name.

File: JabbR/Services/ChatService.cs [hunk 3]
```diff
@@ -130,12 +130,19 @@ public ChatRoom AddRoom(ChatUser user, string name)
...
+        public void JoinRoom(ChatUser user, ChatRoom room, string inviteCode)
         {
             // Throw if the room is private but the user isn't allowed
-            if (room.Private && !IsUserAllowed(room, user))
+            if (room.Private)
             {
-                throw new InvalidOperationException(String.Format("Unable to join {0}. This room is locked and you don't have permission to enter.", room.Name));
+                if (!String.IsNullOrEmpty(inviteCode) && String.Equals(inviteCode, room.InviteCode, StringComparison.OrdinalIgnoreCase))
+                {
+                    room.AllowedUsers.Add(user);
```
Comment (davidfowl) [inline JabbR/Services/ChatService.cs:142]: Add a comment here for this new behavior.

Comment (analogrelay) [inline JabbR/Services/ChatService.cs:142]: Done

File: JabbR/Commands/CommandManager.cs (hunk 2 above)
Comment (analogrelay) [inline JabbR/Commands/CommandManager.cs]: Done: Infrastructure\RandomUtils

File: JabbR/Chat.js (hunk 1 above)
Comment (analogrelay) [inline JabbR/Chat.js]: Fixed

**Pull request merged**